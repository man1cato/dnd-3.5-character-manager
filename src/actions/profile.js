import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
