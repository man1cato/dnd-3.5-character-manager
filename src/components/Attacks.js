import React from 'react';

const Attacks = ({attacks, onInputChange}) => (        
    <div className=" grid grid--6col">
        <h5 className="grid__col1">Attack</h5>
        <h5>Base</h5>
        <div></div>
        <h5>Mod</h5>
        <div></div>
        <h5>Total</h5>

        <div className="grid__col1">Melee</div>
        <div>{attacks.melee.base}</div>
        <div>+</div>
        <input
            type="text"
            name="attacks"
            id="melee"
            value={attacks.melee.mod}
            onChange={onInputChange}
        />
        <div>=</div>
        <div>{attacks.melee.total}</div>

        <div className="grid__col1">Ranged</div>
        <div>{attacks.ranged.base}</div>
        <div>+</div>
        <input
            type="text"
            name="attacks"
            id="ranged"
            value={attacks.ranged.mod}
            onChange={onInputChange}
        />
        <div>=</div>
        <div>{attacks.ranged.total}</div>

        <div className="grid__col1">Grapple</div>
        <div>{attacks.grapple.base}</div>
        <div>+</div>
        <input
            type="text"
            name="attacks"
            id="grapple"
            value={attacks.grapple.mod}
            onChange={onInputChange}
        />
        <div>=</div>
        <div>{attacks.grapple.total}</div>
    </div>
)

export default Attacks