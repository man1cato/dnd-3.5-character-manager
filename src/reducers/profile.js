export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
