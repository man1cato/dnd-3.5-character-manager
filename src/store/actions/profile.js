import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
