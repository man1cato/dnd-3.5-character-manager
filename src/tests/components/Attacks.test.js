import React from 'react';
import { shallow } from 'enzyme';

import Attacks from '../../components/Attacks';
import profile from '../fixtures/profile';

const attacks = {
    bab: profile.fields.attacks.bab,
    melee: {
        base: profile.fields.attacks.melee.base,
        mod: 3
    },
    ranged: {
        base: profile.fields.attacks.ranged.base,
        mod: 3
    },
    grapple: {
        base: profile.fields.attacks.grapple.base,
        mod: 3
    }
}

test('should render attacks with profile data', () => {
    const wrapper = shallow(<Attacks attacks={attacks} />);
    expect(wrapper).toMatchSnapshot();
})
