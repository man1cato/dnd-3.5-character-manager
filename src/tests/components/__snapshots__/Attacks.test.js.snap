// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render attacks with profile data 1`] = `
<div
  className=" grid grid--6col"
>
  <h5
    className="grid__col1"
  >
    Attack
  </h5>
  <h5>
    Base
  </h5>
  <div />
  <h5>
    Mod
  </h5>
  <div />
  <h5>
    Total
  </h5>
  <div
    className="grid__col1"
  >
    Melee
  </div>
  <div>
    4
  </div>
  <div>
    +
  </div>
  <input
    id="melee"
    name="attacks"
    type="text"
    value={3}
  />
  <div>
    =
  </div>
  <div />
  <div
    className="grid__col1"
  >
    Ranged
  </div>
  <div>
    6
  </div>
  <div>
    +
  </div>
  <input
    id="ranged"
    name="attacks"
    type="text"
    value={3}
  />
  <div>
    =
  </div>
  <div />
  <div
    className="grid__col1"
  >
    Grapple
  </div>
  <div>
    4
  </div>
  <div>
    +
  </div>
  <input
    id="grapple"
    name="attacks"
    type="text"
    value={3}
  />
  <div>
    =
  </div>
  <div />
</div>
`;
